cmd/fts/cmd/task.go

```diff
@@ -0,0 +1,117 @@
+package cmd
+
+import (
+	"sync"
+	"time"
+
+	client "github.com/textileio/powergate/api/client"
+	userPb "github.com/textileio/powergate/api/gen/powergate/user/v1"
+)
+
+// Task is an individual data collection to be stored on Filecoin
+type Task struct {
+	Name          string               `json:"name,omitempty"`
+	Path          string               `json:"path,omitempty"`
+	Bytes         int64                `json:"bytes,omitempty"`
+	IsDir         bool                 `json:"isDir,omitempty"`
+	IsHidden      bool                 `json:"isHidden,omitempty"`
+	CID           string               `json:"cid,omitempty"`
+	JobID         string               `json:"jobID,omitempty"`
+	Stage         TaskStage            `json:"stage,omitempty"`
+	Error         string               `json:"error,omitempty"`
+	Deals         []*userPb.StorageJob `json:"deals,omitempty"`
+	err           error
+	storageConfig *userPb.StorageConfig
+}
+
+type queueConfig struct {
+	watch    chan Task
+	updates  chan<- Task
+	complete chan struct{}
+	staging  *StagingStatus
+	pipeWg   *sync.WaitGroup
+	pipe     PipelineConfig
+}
+
+func queue(task Task, conf queueConfig, isLast bool) {
+	for {
+		// check available space
+		ready, err := conf.staging.Ready(task.Bytes)
+		waitTimeout := 1 * time.Second
+		if err != nil {
+			task.Error = err.Error()
+			task.err = err
+			conf.updates <- task
+			conf.pipeWg.Done()
+			return
+		}
+		// wait for space
+		if !ready {
+			time.Sleep(waitTimeout)
+			waitTimeout = waitTimeout * 2
+			continue
+		}
+		// start processing
+		conf.pipeWg.Add(1)
+		go process(task, conf, isLast)
+		return
+	}
+}
+
+func process(task Task, conf queueConfig, isLast bool) {
+	defer func() {
+		if isLast {
+			conf.complete <- struct{}{}
```

Comment:
fires a complete message to the Job Watcher channel, which is a single channel that polls results. This complete message doesn't tell it to stop, just that it can expect to stop getting results soon.